corehq/apps/linked_domain/README.md

```diff
@@ -33,20 +33,30 @@ Linked apps can be setup between linked domains by running `link_app_to_remote`
 
 # Linked Applications
 
-Linked applications predate linked domains. Now that linked domains exist, when you link an app, the linked domain record is automatically created.
+Linked applications predate linked domains. Now that linked domains exist, when you link an app, the linked domain record is automatically created. A linked/downstream app is tied to one or more master/upstream apps via the `upstream_app_id` and `upstream_version` attributes. 
 
 ## Pulling changes from master
-A linked app can be pulled if its master app has a higher released version than the current version of the linked app. Pulling a linked app is similar but not identical to copying an app.
+A linked app can be pulled if its master/upstream app has a higher released version than the `upstream_version` of the linked app. Pulling a linked app is similar but not identical to copying an app.
 
 When a linked/downstream app is pulled from its master/upstream app:
 - The linked app's version will be incremented.
 - The two apps will have **different** ids.
 - Corresponding modules in the master and linked app will have the **same** unique ids.
-- Corresponding forms in the master and linked app will have **different** unique ids.
 - Corresponding forms in the master and linked app will have the **same** XMLNS.
+- Corresponding forms in the master and linked app will have **different** unique ids.
+   - In order to keep these ids consistent from one pull to another, particularly when using multiple master apps (see below), part of the pull process maps the forms in the new linked app to the forms in the most recent version of the linked app. This overwrites the form ids copied from the master app with the form ids that are being used in the linked app. Any new forms (added to the master since the last time the linked app was pulled) will be assigned brand-new ids.
+   - Note on shadow forms: The mapping described above is based on XMLNS. Since a shadow form inherits its parent's XMLNS, shadow forms cannot be mapped. The result of this is that any shadow forms in the linked app are unstable: their unique ids change on **every** pull. This means that workflows that store a form's unique id, such as UCRs based on a form data source, will not work with shadow forms on linked apps.
```

Comment:
:+1:

Reply:
might be useful to put "The mapping described above is based on XMLNS" in the para above.
"This overwrites the form ids copied from the master app with the form ids that are being used in the linked app **by matching the XMLNS**"

Reply:
@esoergel 
so "Note on shadow forms" was always true and was noticed now?
What is the implication of that?

Reply:
yeah, that was always true.  It means that identifying shadow forms by unique ID in a linked app is unlikely to work.  Could have ramifications for reporting, I suppose, though I think that usually uses xmlns.